drivers/sdhc/sdhc_esp32.c

```diff
@@ -0,0 +1,1544 @@
+static bool wait_for_busy_cleared(uint32_t timeout_ms)
+{
+	if (timeout_ms == 0) {
+		return !sdmmc_host_card_busy();
+	}
+
+	/* It would have been nice to do this without polling, however the peripheral
+	 * can only generate Busy Clear Interrupt for data write commands, and waiting
+	 * for busy clear is mostly needed for other commands such as MMC_SWITCH.
+	 */
+	uint32_t timeout_ticks = k_ms_to_ticks_ceil32(timeout_ms);
+
+	while (timeout_ticks-- > 0) {
+		if (!sdmmc_host_card_busy()) {
+			return true;
+		}
+		k_sleep(K_MSEC(1));
+	}
+
+	return false;
+}
+
+static bool cmd_needs_auto_stop(const struct sdmmc_command *cmd)
+{
+	/* SDMMC host needs an "auto stop" flag for the following commands: */
+	return cmd->datalen > 0 &&
+	       (cmd->opcode == MMC_WRITE_BLOCK_MULTIPLE || cmd->opcode == MMC_READ_BLOCK_MULTIPLE ||
+		cmd->opcode == MMC_WRITE_DAT_UNTIL_STOP || cmd->opcode == MMC_READ_DAT_UNTIL_STOP);
+}
+
+static sdmmc_hw_cmd_t make_hw_cmd(struct sdmmc_command *cmd)
+{
+	sdmmc_hw_cmd_t res = {0};
+
+	res.cmd_index = cmd->opcode;
+	if (cmd->opcode == MMC_STOP_TRANSMISSION) {
+		res.stop_abort_cmd = 1;
+	} else if (cmd->opcode == MMC_GO_IDLE_STATE) {
+		res.send_init = 1;
+	} else {
+		res.wait_complete = 1;
+	}
+	if (cmd->opcode == MMC_GO_IDLE_STATE) {
+		res.send_init = 1;
+	}
+	if (cmd->flags & SCF_RSP_PRESENT) {
+		res.response_expect = 1;
+		if (cmd->flags & SCF_RSP_136) {
+			res.response_long = 1;
+		}
+	}
+	if (cmd->flags & SCF_RSP_CRC) {
+		res.check_response_crc = 1;
+	}
+	if (cmd->data) {
+		res.data_expected = 1;
+
+		if ((cmd->flags & SCF_CMD_READ) == 0) {
+			res.rw = 1;
+		}
+
+		if ((cmd->datalen % cmd->blklen) != 0) {
+			return res; /* Error situation, data will be invalid */
+		}
+
+		res.send_auto_stop = cmd_needs_auto_stop(cmd) ? 1 : 0;
+	}
+	LOG_DBG("%s: opcode=%d, rexp=%d, crc=%d, auto_stop=%d", __func__, res.cmd_index,
+		res.response_expect, res.check_response_crc, res.send_auto_stop);
+
+	return res;
+}
+
+static int sdmmc_host_do_transaction(int slot, struct sdmmc_command *cmdinfo)
+{
+	int ret;
+
+	if (k_mutex_lock(&s_request_mutex, K_FOREVER) != 0) {
+		return ESP_ERR_NO_MEM;
+	}
+
+	/* dispose of any events which happened asynchronously */
+	handle_idle_state_events();
+
+	/* convert cmdinfo to hardware register value */
+	sdmmc_hw_cmd_t hw_cmd = make_hw_cmd(cmdinfo);
+
+	if (cmdinfo->data) {
+		/* Length should be either <4 or >=4 and =0 (mod 4) */
+		if ((cmdinfo->datalen >= 4) && (cmdinfo->datalen % 4) != 0) {
+			LOG_DBG("%s: invalid size: total=%d", __func__, cmdinfo->datalen);
+			ret = ESP_ERR_INVALID_SIZE;
+			goto out;
+		}
+
+		if ((((intptr_t)cmdinfo->data % 4) != 0) || !esp_ptr_dma_capable(cmdinfo->data)) {
+			LOG_DBG("%s: buffer %p can not be used for DMA", __func__, cmdinfo->data);
+			ret = ESP_ERR_INVALID_ARG;
+			goto out;
+		}
+
+		/* this clears "owned by IDMAC" bits */
+		memset(s_dma_desc, 0, sizeof(s_dma_desc));
+
+		/* initialize first descriptor */
+		s_dma_desc[0].first_descriptor = 1;
```

Comment:
We should not be using static global variables here. The DMA descriptors (and other static global variables) should be in the device instance data structure. This way, we can support multiple instances of the SDHC controller in one build image, provided there are multiple on one SOC